### PR TITLE
[DI] Allow explicit executor for torchscript interpreter

### PR DESCRIPTION
**Pull request by garroud**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#46444 [DI] Allow explicit executor for torchscript interpreter**

add explicit executor for torchscript interpreter.

Differential Revision: [D24114874](https://our.internmc.facebook.com/intern/diff/D24114874/)

### CODE REVIEW

File: test/cpp/jit/test_interpreter.cpp [hunk 1]
```diff
@@ -138,5 +142,38 @@ TEST(InterpreterTest, Basic_CUDA) {
...
+  The test module file is generated by following:
+    class DemoModule(torch.nn.Module):
+      def forward(self):
+        r1 = torch.jit.fork(torch.mm, torch.rand(100,100),torch.rand(100,100))
+        r2 = torch.jit.fork(torch.mm, torch.rand(100,100),torch.rand(100,100))
+        return r1.wait() + r2.wait()
+  demo = DemoModule()
+  torch.jit.save(torch.jit.script(demo), 'test_interpreter_async.pth')
+  */
+  auto model = load("caffe2/test/cpp/jit/test_interpreter_async.pth");
```
Comment (dzhulgakov) [inline test/cpp/jit/test_interpreter.cpp:159]: this won't work in OSS CI - you need to change it to a relative path (or wait for the parser problem to be fixed)

examples:
https://github.com/pytorch/pytorch/blob/master/test/jit/test_save_load.py#L13
https://github.com/pytorch/pytorch/blob/master/test/quantization/test_backward_compatibility.py#L29

Comment (houseroad) [reply test/cpp/jit/test_interpreter.cpp:159]: I thought we are able to load the models by copying the generated code, so we are still not able to do so?

Comment (garroud) [reply test/cpp/jit/test_interpreter.cpp:159]: forks seems to have problem with the parser.

**Comment by garroud** [comment]
Do we allow to use Boost here? the build failure seems due to I did not add Boost into Cmake

**Review by houseroad — COMMENTED** [review]
Boost is not dependency of PyTorch, and I don't think it's due to Boost ( I didn't see any dependency on Boost).

Related error message: 
```
test_interpreter.cpp:(.text+0x2b7e): undefined reference to `torch::jit::InterpreterState::runAsync(std::vector<c10::IValue, std::allocator<c10::IValue> >&)'
```
You probably need to expose `InterpreterState` as well.

**Review by houseroad — COMMENTED** [review]
Test failures in backward compatibility check is unrelated, please rebase your diff and reexport the diff to github.

pytorch_linux_bionic_py3_6_clang9_test's failure is related, seems the model fails to load.

**Review by houseroad — COMMENTED** [review]
Also please use .pt extension instead .pth.

**Comment by garroud** [comment]
hgexport not work with binary file. close this PR.

**Pull request closed without merging**